### PR TITLE
Fix #71677

### LINKED ISSUE #71677: Array join with a nested data structure, which contains an alias to a constant array leads to a null pointer dereference
```
CREATE TABLE test
(
    x UInt8,
    “struct.x” DEFAULT [0],
    “struct.y” ALIAS [1],
)
ENGINE = Memory;

insert into test (x) values (0);
select * from test array join struct;
```

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Array join with a nested data structure, which contains an alias to a constant array was leading to a null pointer dereference. This closes #71677.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

File: src/Functions/nested.cpp [hunk 1]
```diff
@@ -108,27 +108,29 @@ class FunctionNested : public IFunction
     {
         size_t arguments_size = arguments.size();
 
-        const auto * lhs_array = assert_cast<const ColumnArray *>(arguments.at(1).column.get());
```
Comment (alexey-milovidov) [inline src/Functions/nested.cpp:111]: `auto` is a bad culture.
Pointers are a bad culture.

File: src/Functions/nested.cpp [hunk 2]
```diff
@@ -168,7 +170,12 @@ REGISTER_FUNCTION(Nested)
 {
     factory.registerFunction<FunctionNested>(FunctionDocumentation{
         .description=R"(
+This is a function used internally by the ClickHouse engine and not meant to be used directly.
```
Comment (alexey-milovidov) [inline src/Functions/nested.cpp:173]: The documentation was also bad.

File: src/Functions/nested.cpp [hunk 3]
```diff
@@ -108,27 +108,29 @@ class FunctionNested : public IFunction
...
+            const ColumnArray & other_array = assert_cast<const ColumnArray &>(*other_array_materialized);
 
-            if (!lhs_array->hasEqualOffsets(*rhs_array))
+            if (!first_array.hasEqualOffsets(other_array))
                 throw Exception(ErrorCodes::SIZES_OF_ARRAYS_DONT_MATCH,
                     "The argument 2 and argument {} of function {} have different array offsets",
                     i + 1,
                     getName());
 
-            data_columns.push_back(rhs_array->getDataPtr());
```
Comment (alexey-milovidov) [inline src/Functions/nested.cpp:127]: Variable names were misleading.

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 19ca58e95203ed2eed71a6ef0ab88677d9bb6b93 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71678/19ca58e95203ed2eed71a6ef0ab88677d9bb6b93/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by cwurm** [comment]
Can we backport this?

**Comment by Algunenano** [comment]
Backport to which version? 24.8 I guess? I don't see a problem as the changes look minimal

**Pull request merged**